networkdb/nodemgmt.go

```diff
@@ -0,0 +1,120 @@
+package networkdb
+
+import (
+	"fmt"
+
+	"github.com/hashicorp/memberlist"
+	"github.com/sirupsen/logrus"
+)
+
+type nodeState int
+
+const (
+	nodeNotFound    nodeState = -1
+	nodeActiveState nodeState = 0
+	nodeLeftState   nodeState = 1
+	nodeFailedState nodeState = 2
+)
+
+var nodeStateName = map[nodeState]string{
+	-1: "NodeNotFound",
+	0:  "NodeActive",
+	1:  "NodeLeft",
+	2:  "NodeFailed",
+}
+
+// findNode search the node into the 3 node lists and returns the node pointer and the list
+// where it got found
+func (nDB *NetworkDB) findNode(nodeName string) (*node, nodeState, map[string]*node) {
+	for i, nodes := range []map[string]*node{
+		nDB.nodes,
+		nDB.leftNodes,
+		nDB.failedNodes,
+	} {
+		if n, ok := nodes[nodeName]; ok {
+			return n, nodeState(i), nodes
+		}
+	}
+	return nil, nodeNotFound, nil
+}
+
+// changeNodeState changes the state of the node specified, returns true if the node was moved,
+// false if there was no need to change the node state. Error will be returned if the node does not
+// exists
+func (nDB *NetworkDB) changeNodeState(nodeName string, newState nodeState) (bool, error) {
+	n, currState, m := nDB.findNode(nodeName)
+	if n == nil {
+		return false, fmt.Errorf("node %s not found", nodeName)
+	}
+
+	switch newState {
+	case nodeActiveState:
+		if currState == nodeActiveState {
+			return false, nil
+		}
+
+		delete(m, nodeName)
+		// reset the node reap time
+		n.reapTime = 0
+		nDB.nodes[nodeName] = n
+	case nodeLeftState:
+		if currState == nodeLeftState {
+			return false, nil
+		}
+
+		delete(m, nodeName)
+		nDB.leftNodes[nodeName] = n
+	case nodeFailedState:
+		if currState == nodeFailedState {
+			return false, nil
+		}
+
+		delete(m, nodeName)
+		nDB.failedNodes[nodeName] = n
+	}
+
+	logrus.Infof("Node %s change state %s --> %s", nodeName, nodeStateName[currState], nodeStateName[newState])
+
+	if newState == nodeLeftState || newState == nodeFailedState {
+		// set the node reap time, if not already set
+		// It is possible that a node passes from failed to left and the reaptime was already set so keep that value
+		if n.reapTime == 0 {
+			n.reapTime = nodeReapInterval
+		}
+		// The node leave or fails, delete all the entries created by it.
+		// If the node was temporary down, deleting the entries will guarantee that the CREATE events will be accepted
+		// If the node instead left because was going down, then it makes sense to just delete all its state
+		nDB.deleteNodeFromNetworks(n.Name)
+		nDB.deleteNodeTableEntries(n.Name)
+	}
+
+	return true, nil
+}
+
+func (nDB *NetworkDB) purgeReincarnation(mn *memberlist.Node) bool {
```

Comment:
General comment on this function. In the previous incarnation of this function, the nodes were purged if the address matches. But now, I see 2 changes
1. The nodes are also purged from Active state (which wasnt the case before)
2. The nodes are purged only if the Port matches and the Name doesnt match.
These are obviously change in functionalities than just reorganizing the functions.

Can you pls give more clarifications on these ?

Reply:
replying to the points:
1) this function is now called only during the notifyJoin from memberlist. The check for active is to cover the case of a quick node reboot, where a new instance from the same host (same ip, and same port but new name that is autogenerated) is joining the cluster, before that the previous incarnation is marked as dead from memberlist because of the ping failure. In this specific case this function identify that the node is actually a new incarnation of the same node so will take care of the cleanup of the previous one before the arrival of the memberlist notification. The main objective here is to avoid that a node with same IP is sent into the failed list, if doing so the node will be polled periodically from the retry logic, but in this case is the wrong decision because that identity is gone forever and replaced by a new one.

2) The condition for a new incarnation is to have same IP, same port but different name. If a node joins with same name and same IP is not a new incarnation is simply one node that got disconnected and came back and anyway won't arrive here, but will be intercepted by the logic at line https://github.com/docker/libnetwork/pull/2035/files#diff-e0a0ef8c8c9573b1dabd65379df2af3fL51 and following.

Overall I though that would be clearer to have 1 single function that takes care of all the conditions and address them on the join on a node instead of splitting the logic between join/leave